bugsnag-android-core/src/androidTest/java/com/bugsnag/android/JournalTest.kt

```diff
@@ -0,0 +1,124 @@
+package com.bugsnag.android
+
+import com.bugsnag.android.internal.Journal
+import junit.framework.TestCase.assertNotNull
+import junit.framework.TestCase.assertNull
+import org.junit.Assert
+import org.junit.Test
+import java.io.ByteArrayOutputStream
+
+class JournalTest {
+    @Test
+    fun testEmptyEverything() {
+        val expected = mutableMapOf<String, Any>()
+        val journal = Journal()
+        val document = mutableMapOf<String, Any>()
+        val observed = journal.applyTo(document)
+        Assert.assertEquals(expected, observed)
+    }
+
+    @Test
+    fun testBasicMap() {
+        val expected = mutableMapOf<String, Any>("a" to 1)
+        val journal = Journal()
+        journal.add(Journal.Command("a", 1))
+        val document = mutableMapOf<String, Any>()
+        val observed = journal.applyTo(document)
+        Assert.assertEquals(expected, observed)
+    }
+
+    @Test
+    fun test1EntryJournal() {
+        assertSerializeDeserialize(
+            "{\"a\":1}\u0000".toByteArray(Charsets.UTF_8),
+            Journal.Command("a", 1)
+        )
+    }
+
+    @Test
+    fun test2EntryJournal() {
+        assertSerializeDeserialize(
+            "{\"a\":1}\u0000{\"b\":\"x\"}\u0000".toByteArray(Charsets.UTF_8),
+            Journal.Command("a", 1),
+            Journal.Command("b", "x")
+        )
+    }
+
+    @Test
+    fun testRunSerializedJournal() {
+        assertRunSerializedJournal(
+            mutableMapOf("q" to 100),
+            "{\"a\":1}\u0000{\"b\":\"x\"}\u0000".toByteArray(Charsets.UTF_8),
+            mutableMapOf("q" to 100, "a" to 1, "b" to "x")
+        )
+    }
+
+    @Test
+    fun testRunSerializedJournalDeep() {
+        assertRunSerializedJournal(
+            mutableMapOf("q" to 100),
+            "{\"a.b.-1\":1}\u0000{\"b.-1.\":\"x\"}\u0000".toByteArray(Charsets.UTF_8),
+            mutableMapOf(
+                "q" to 100,
+                "a" to mapOf<String, Any>("b" to listOf<Any>(1.toLong())),
+                "b" to listOf<Any>(listOf<Any>("x"))
+            )
+        )
+    }
+
+    @Test
+    fun testDeserializeFail() {
+        assertDeserializeFails(
+            "{\"a.\":10false}\u0000".toByteArray(Charsets.UTF_8)
+        )
+    }
+
+    private fun assertDeserializeFails(serialized: ByteArray) {
+        val logger = InterceptingLogger()
+        assertNull(logger.msg)
+        deserialize(serialized, logger)
+        assertNotNull(logger.msg)
```

Comment:
Is feels a bit like overfitting though. The deserialize routine is only supposed to write to the log if an error occurs, and never write to the log if everything decodes successfully. The exact log contents aren't important other than that they should adequately describe the problems encountered so that we can report them somewhere (if we care to).

Reply:
Thinking more about this it feels a little problematic that `deserialize()` can return an incompletely deserialized `Journal` object and the caller won't have any idea of what went wrong. Perhaps it'd be worth throwing an exception instead, or invoking some callback failure, so that it can be handled upstream?

Reply:
Journal now throws an exception as soon as it encounters a deserialization problem.